### DISCUSSION TITLE
does it support eslint and formatter?

### DISCUSSION THREAD

**Discussion by expoure** [helix-editor/helix#1930] [Q&A]
i know there's lsp support, but what about eslint and prettier?

**Comment by sudormrfbin** [comment]
Formatting is done through the LSP right now, so if the LSP has a configurable way of setting prettier to be the formatter it'll work (I don't do much web dev so I'm not sure about it).

**Reply by expoure** [reply]
I guess there is not a way to setting it (for while). I may be wrong, so if someone know how to set eslint and prettier (and others linters and formatters) i will be thanks for that.

**Reply by David-Else** [reply]
Eslint has an LSP! There is a package that contains all the up-to-date vs code language servers:

-  vscode-html-language-server
-  vscode-css-language-server
-  vscode-json-language-server
-  vscode-eslint-language-server <<<<<<<<<< :)

https://www.npmjs.com/package/vscode-langservers-extracted

I am hoping this gets used as default https://github.com/helix-editor/helix/issues/1895

BUT we need to be able to use 2x LSPs as we still need the TypeScript Language server :( https://github.com/helix-editor/helix/issues/1396

I would love to get prettier working, and have a way to turn off formatting in certain language servers. Sometimes prettier is better.

I read that you can pipe the contents of a buffer through the command line, so it should not be hard to use Prettier that way. In vim it sucks doing that as the cursor goes back to the top of the document, I am not sure what would happen in Helix

There is a rust crate for `dprint` that can be used as a drop in replacement for Prettier in some cases, https://dprint.dev/ , it is used in Deno.

**Reply by David-Else** [reply]
You can also use Deno as the LSP and get the full language server, formatting and linting combined, but it is not an out-of-the-box setup as you will have the problem that imports require the `.ts` extension and that will throw a spanner in the works for a lot of things. You can bypass that problem by building with `esbuild` deno extension, but that is something else.

**Reply by expoure** [reply]
> Eslint has an LSP! There is a package that contains all the up-to-date vs code language servers:
> 
> * vscode-html-language-server
> * vscode-css-language-server
> * vscode-json-language-server
> * vscode-eslint-language-server <<<<<<<<<< :)
> 
> https://www.npmjs.com/package/vscode-langservers-extracted
> 
> I am hoping this gets used as default #1895
> 
> I would love to get prettier working, and have a way to turn off formatting in certain language servers. Sometimes prettier is better.
> 
> I read that you can pipe the contents of a buffer through the command line, so it should not be hard to use Prettier that way. In vim it sucks doing that as the cursor goes back to the top of the document, I am not sure what would happen in Helix
> 
> There is a rust crate for `dprint` that can be used as a drop in replacement for Prettier in some cases, https://dprint.dev/ , it is used in Deno.

i mean there is no way to setup this in helix right now, no?

**Reply by David-Else** [reply]
I added to my answer, but I think we need https://github.com/helix-editor/helix/issues/1396 @sudormrfbin will know, I only started using helix yesterday!

**Reply by DrSensor** [reply]
I just tried `deno lsp` and seems like there is long standing issue for that https://github.com/helix-editor/helix/issues/1000 (based on the log, seems helix add extra args `--stdio` which not what deno expect). So the best bet is to use either:
* dprint or prettier via pipe-shell, or
* add vscode-eslint-language-server into languages.toml

**Reply by DrSensor** [reply]
Just a quick note, typescript-language-server has a builtin formatter but you need to add [`tsfmt.json`](https://github.com/microsoft/vscode/blob/main/tsfmt.json) in your root project.
```toml
# languages.toml [[language]] name = "typescript"
roots = ["tsfmt.json"] # select one of this ["tsfmt.json"], ["jsconfig.json"], ["package.json"], ["tsconfig.json"], ["deno.json"]
language-server = { command = "typescript-language-server", args = ["--stdio"], language-id = "typescript" }
indent = { tab-width = 2, unit = "  " }
```

Full tsfmt.json settings: https://github.com/microsoft/TypeScript/blob/4d2fb5407c4e9999dd108972290fc57d3dd56e71/lib/protocol.d.ts#L2587-L2615

**Reply by chrisdavies** [reply]
Swank. Without the `tsfmt.json` file, Helix was removing *all* indentation when I saved. :/ Not sure why that was.

**Reply by kirawi** [reply]
Were you using master or `22.03`? There was a bug that was fixed in https://github.com/helix-editor/helix/commit/c18de0e8f001041e234b2b9bb0f8fea122858ad3

**Reply by chrisdavies** [reply]
`22.03`. Is master recommended? I didn't RTFM, so that's my bad, if so.

**Reply by kirawi** [reply]
It's up to you, but in my experience master is very stable.

**Comment by cakir-enes** [comment]
Any updates or workarounds for this? This keeps me away switching from nvim :G

**Reply by cem-al** [reply]
I agree, no support for Eslint is making it hard to use this at work.

**Reply by kirawi** [reply]
https://github.com/helix-editor/helix/pull/2507
